### DISCUSSION TITLE
Example in tutorial for modeling storage

### DISCUSSION THREAD

**Discussion by zsalimian** [iiasa/message_ix#534] [MESSAGE modeling]
Dear Sir,

Is it possible to add an example about modeling storages in tutorial?

Regards

**Comment by LauWien** [comment]
Dear @zsalimian,
Since this is neither about (a) a bug or error in the MESSAGEix package nor (b) a requested feature in the software per se, I have converted this from an issue to a “Discussion”. As the documentation indicates, please only open issues if (a) or (b) is the case; otherwise use the Discussions feature.

**Comment by behnam-zakeri** [comment]
Thank you @zsalimian for the suggestion. We will develop a tutorial for showing the functionality of storage and its features in the coming months. Meanwhile, if you are curious to see how you can set up a storage solution, you can refer to some [simple examples ](https://github.com/iiasa/message_ix/blob/main/message_ix/tests/test_feature_storage.py) that we use for testing. Please kindly note that these testing examples are not meant to teach the users how to add storage. If you find these examples not self-explanatory or useful, you need to wait until we publish a dedicated tutorial for storage.

**Reply by zsalimian** [reply]
thanks for answer. Would you please clarify the role of cooling and dummay commodities in this example?
Is it necessary to define them?
If we do not consider these two commodities, how we should define the input and output for dam (storage device)?

Regards

**Comment by behnam-zakeri** [comment]
This is an optional feature. The role of "cooling" and "dummay" is to link respectively input and output commodities to the storage device. This can be useful in a special case, when the user wants to model a storage device with an input commodity for running the device. Please note that this commodity for running the device is different from the commodity that the storage device can store. For example, a heat storage device can store heat (thermal energy), and the user can link electricity into this device for operating that and keeping that warm over time. Or another example can be a refrigerator for storing cold products, and the user can link electricity to the device for keeping the stored material cool over time (here the stored commodity can be water, etc.). The use of this input commodity is not related to the efficiency of the device, but rather for the duration of storage: the longer storage time, the more use of the input commodity.
If you don't need to connect an input commodity to the storage container to keep it operating, you can remove such commodities ("cooling" and "dummay") from your example. Storage container will work without any input commodity as well, and you can specify a self-discharge, charging or discharging losses through other parameters.

**Reply by zsalimian** [reply]
Many thanks,

It helped a lot. 
There is an ambiguous regarding the results when adding upper bound=0.25 on activity of wind_ppl. The demand in different seasons in 2020 is as follow:
{"a": 0.15, "b": 0.2, "c": 0.4, "d": 0.25}
But when imposing bound activity up=0.25 the output of wind in time sections will reach 0.25!we don't necessarily need such output in "a" or "b" time sections.

   node_loc technology  year_vtg  year_act  mode time   lvl   mrg
0  Westeros   wind_ppl      2020      2020  mode    a  0.25   0.0
1  Westeros   wind_ppl      2020      2020  mode    b  0.25   0.0
2  Westeros   wind_ppl      2020      2020  mode    c  0.25   0.0
3  Westeros   wind_ppl      2020      2020  mode    d  0.25   0.0
4  Westeros    gas_ppl      2020      2020  mode    a  0.00  20.0
5  Westeros    gas_ppl      2020      2020  mode    b  0.00  20.0
6  Westeros    gas_ppl      2020      2020  mode    c  0.15   0.0
7  Westeros    gas_ppl      2020      2020  mode    d  0.00  20.0

Moreover, in addition to the above-mentioned problem, the storage depletion is another issue. You did not impose any constraint that it is a "must" for storage to deplete at the end of period. But turbine produce 0.08 in "a" time section, which is not necessarily required as the demand in this time section was satisfied by wind_ppl.

0      node   wind_ppl      2020      2020  mode    a  0.25   0.000000e+00
1      node   wind_ppl      2020      2020  mode    b  0.25   0.000000e+00
2      node   wind_ppl      2020      2020  mode    c  0.25   0.000000e+00
3      node   wind_ppl      2020      2020  mode    d  0.25   0.000000e+00
4      node    gas_ppl      2020      2020  mode    a  0.00   2.000000e+00
5      node    gas_ppl      2020      2020  mode    b  0.00   2.000000e+00
6      node    gas_ppl      2020      2020  mode    c  0.15   0.000000e+00
7      node    gas_ppl      2020      2020  mode    d  0.00   0.000000e+00
8      node        dam      2020      2020  mode    a  0.00   0.000000e+00
9      node        dam      2020      2020  mode    b  0.00   0.000000e+00
10     node        dam      2020      2020  mode    c  0.00  4.940656e-324
11     node        dam      2020      2020  mode    d  0.00  4.940656e-324
12     node       pump      2020      2020  mode    a  0.00   5.000000e-01
13     node       pump      2020      2020  mode    b  0.00   5.000000e-01
14     node       pump      2020      2020  mode    c  0.00   5.000000e-01
15     node       pump      2020      2020  mode    d  0.00   5.000000e-01
16     node    turbine      2020      2020  mode    a  0.08   0.000000e+00
17     node    turbine      2020      2020  mode    b  0.00   0.000000e+00
18     node    turbine      2020      2020  mode    c  0.00   0.000000e+00
19     node    turbine      2020      2020  mode    d  0.00   0.000000e+00, 


Regards

**Comment by behnam-zakeri** [comment]
Thanks. Related to your questions:
1. wind generates more than demand in time slice "a", and this surplus cheap electricity is stored in the storage device and is discharged in time slice "c", when there is high demand and low wind. This is the functionality of storage: shifting electricity from one time slice to another to reduce the costs.
2. In the model results, for scenario "with-storage-equal-time", the storage is charged in time slice "a" and "b", and discharged in time slice "c". From my GDX file, I see that turbine is only active in time slice "c". Please see the snapshot from the output GDX below. Maybe I didn't understand your question.
![image](https://user-images.githubusercontent.com/30926636/143881694-4f1498b5-739c-44c0-95a5-a56f2d6f6329.png)

**Reply by zsalimian** [reply]
thank you for your reply,
You are talking about the fourth case, where dam has input and output. But it is not my problem.
I am talking about the first 3 cases.
1-In the first case, we have demand={"a": 0.15, "b": 0.2, "c": 0.4, "d": 0.25} and the answer exactly satisfies this demand.
2-In the second case, we have demand={"a": 0.15, "b": 0.2, "c": 0.4, "d": 0.25} and we impose upper boundary on activity**_. There is no storage_** but the optimal result show that wind_ppl should produce 0.25 in "a" and "b" time sections, where we don't need that amount, and of course there is no storage to store excess supply.
3- In the third case, we have storage without any input. In this case, the optimal result show that wind_ppl should produce 0.25 in "a" and "b" time sections,  but it does not store it in storage. Because the following results show that turbine produce 0.08. Therefor we have storage depletion, but the gas_ppl produce 0.15!


0 node wind_ppl 2020 2020 mode a 0.25 0.000000e+00
1 node wind_ppl 2020 2020 mode b 0.25 0.000000e+00
2 node wind_ppl 2020 2020 mode c 0.25 0.000000e+00
3 node wind_ppl 2020 2020 mode d 0.25 0.000000e+00
4 node gas_ppl 2020 2020 mode a 0.00 2.000000e+00
5 node gas_ppl 2020 2020 mode b 0.00 2.000000e+00
6 node gas_ppl 2020 2020 mode c 0.15 0.000000e+00
7 node gas_ppl 2020 2020 mode d 0.00 0.000000e+00
8 node dam 2020 2020 mode a 0.00 0.000000e+00
9 node dam 2020 2020 mode b 0.00 0.000000e+00
10 node dam 2020 2020 mode c 0.00 4.940656e-324
11 node dam 2020 2020 mode d 0.00 4.940656e-324
12 node pump 2020 2020 mode a 0.00 5.000000e-01
13 node pump 2020 2020 mode b 0.00 5.000000e-01
14 node pump 2020 2020 mode c 0.00 5.000000e-01
15 node pump 2020 2020 mode d 0.00 5.000000e-01
16 node turbine 2020 2020 mode a 0.08 0.000000e+00
17 node turbine 2020 2020 mode b 0.00 0.000000e+00
18 node turbine 2020 2020 mode c 0.00 0.000000e+00
19 node turbine 2020 2020 mode d 0.00 0.000000e+00,

Thank you for your kind reply.

**Reply by zsalimian** [reply]
Finally, I can not understand the role of this syntax in the fourth case.

df = scen.par("output", {"technology": "turbine"})
To explain more,  in the following code you added cooler and cooling. But I can not understand the relation between cooler and turbine, and the role of "scen.par("output", {"technology": "turbine"})" syntax. Because we had added turbine in the third case before.


    # # Adding a new technology "cooler" to provide input of "cooling" to dam
    scen.add_set("technology", "cooler")
    df = scen.par("output", {"technology": "turbine"})
    df["technology"] = "cooler"
    df["commodity"] = "cooling"
    scen.add_par("output", df)


With Best Regards,

**Reply by behnam-zakeri** [reply]
> thank you for your reply, You are talking about the fourth case, where dam has input and output. But it is not my problem. I am talking about the first 3 cases. 1-In the first case, we have demand={"a": 0.15, "b": 0.2, "c": 0.4, "d": 0.25} and the answer exactly ...
> 
Thanks, @zsalimian, for clarifying your question, and for the good points. The answer with the same order:
1. Your interpretation is correct.
2. We see wind generation more than demand because wind power plant is free of charge in this example. You can change the variable cost of "wind_ppl" in line 29 of the code, e.g., to 0.01, then you will see no overproduction from wind. Please note that in MESSAGE, be default, `supply >= demand`, so oversupply in this case as it has zero cost was not a problem.
3. If you add `var_cost` for "wind_ppl" as discussed above, you won't see oversupply from wind. In this example, the storage container ("dam"), needs an input commodity ("cooling") to operate. Because we don't establish a link to any technology that can supply the commodity "cooling" as "input" to storage, the storage solution doesn't work. It means it doesn't charge and store. However, as there is still an initial content in the storage container (specified through parameter `storage_initial`), the discharger depletes that stored commodity as you mentioned.
I hope this calrifies.

**Reply by behnam-zakeri** [reply]
> Finally, I can not understand the role of this syntax in the fourth case.
> df = scen.par("output", {"technology": "turbine"}) To explain more, in the following code you added cooler and cooling. But I can 
> ...

That syntax, i.e., `scenario.par()` can be used to load the data from a parameter in a scenario. Here, we load the data from parameter "output", filtered for "technology" that is "turbine". You can print this DataFrame (looks like a table) and see the content of `df`. Later, we use this table and modify it to create another table for parameter "output" of the technology "cooler". Then, we add the new table to the model through `scen.add_par("output", df)`. There is no relation between "turbine" and "cooler" in the model. We only use the table (DataFrame) from "turbine" as template, to easily generate data for "cooler".

**Reply by zsalimian** [reply]
Many many thanks.
You clarified all.
It helped me a lot to model storages in energy network of my country.
The last question is about testing functionality of storage. 
Is it a "must" to check functionality  using "assert" syntax? The disadvantage of this method is that if the solution does not satisfy the condition it says "error occurred", but does not correct the answer.
Can I define the check conditions as relations in model so that it could affect the optimal solution?

With Best wishes
